sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestPipelineTest.java

```diff
@@ -52,280 +59,491 @@
+        // @NeedsRunner annotation with a CrashingRunner indicates a configuration issue
+      }
 
-    final PCollection<String> pCollection = pCollection();
-    PAssert.that(pCollection).containsInAnyOrder(DUMMY);
-    pipeline.run().waitUntilFinish();
+      @Category(RunnableOnService.class)
+      @Test
+      public void testConflict_runnableOnServiceAnnotation_expectIllegalStateException()
+          throws Exception {
 
-    // dangling PTransform
-    addTransform(pCollection);
-  }
+        addTransform(pCollection(pipeline));
 
-  @Test
-  public void testNormalFlowWithPAssert() throws Throwable {
-    PAssert.that(pCollection()).containsInAnyOrder(DUMMY);
-    pipeline.run().waitUntilFinish();
-  }
+        // @RunnableOnService annotation with a CrashingRunner indicates a configuration issue
+      }
 
-  @Test
-  public void testAutoAddMissingRunFlow() throws Throwable {
-    PAssert.that(pCollection()).containsInAnyOrder(DUMMY);
-    // missing pipeline#run, but have it auto-added.
-    pipeline.enableAutoRunIfMissing(true);
-  }
+    }
+
+    private static List<Object[]> extractTests(final Class<?> testClass) {
+      final ArrayList<org.junit.runner.Description> testDescriptions = Request.aClass(testClass)
+                                                                              .getRunner()
+                                                                              .getDescription()
+                                                                              .getChildren();
+
+      return FluentIterable.from(testDescriptions)
+                           .transform(new Function<org.junit.runner.Description, Object[]>() {
+
+                             @Override
+                             public Object[] apply(final org.junit.runner.Description description) {
+                               return new Object[] {
+                                   Request.method(testClass, description.getMethodName()),
+                                   description.getTestClass().getSimpleName() + "#"
+                                       + description.getMethodName()
+                               };
+                             }
+                           })
+                           .toList();
+    }
+
+    private TestRule withPossibleExceptions(final String testName) {
+      final String expect = testName.substring(testName.indexOf("_expect"));
+      if (expect.contains(TestPipeline.PipelineRunMissingException.class.getSimpleName())) {
+        final ExpectedException exception = ExpectedException.none();
+        exception.expect(TestPipeline.PipelineRunMissingException.class);
+        return exception;
+      } else if (expect.contains(TestPipeline.AbandonedNodeException.class.getSimpleName())) {
+        final ExpectedException exception = ExpectedException.none();
+        exception.expect(TestPipeline.AbandonedNodeException.class);
+        if (testName.contains(P_TRANSFORM)) {
```

Comment:
The expectations set up here can be utility methods, but their calls should be inlined to the test methods themselves. Drive the test with code, not its name. It actually makes it very troublesome to find out what a test is really doing if one needs to go look in these shared conditionals based on string munging.

Instead of:

```java
@Test
@RunnableOnService
public void testConflict_runnableOnServiceAnnotation_expectIllegalStateException()
    throws Exception {
  addTransform(pCollection(pipeline));
  // @RunnableOnService annotation with a CrashingRunner indicates a configuration issue
}		
```

this:

```java
/**
 * A `RunnableOnService` category with a `CrashingRunner` indicates a configuration issue.
 */
@Test
public void testCrashingRunnerRunnableOnService() throws Exception {
  thrown.expect(IllegalStateException.class);
  thrown.expectMessage("some keywords, but not too many, to make the message has pertinent info");
  addTransform(pCollection(pipeline));
}
```